src/contracts/core/AllocationManager.sol

```diff
@@ -0,0 +1,811 @@
+        }
+    }
+
+    /// @inheritdoc IAllocationManager
+    function registerForOperatorSets(
+        address operator,
+        RegisterParams calldata params
+    ) external onlyWhenNotPaused(PAUSED_OPERATOR_SET_REGISTRATION_AND_DEREGISTRATION) checkCanCall(operator) {
+        // Check that the operator exists
+        require(delegation.isOperator(operator), InvalidOperator());
+
+        for (uint256 i = 0; i < params.operatorSetIds.length; i++) {
+            // Check the operator set exists and the operator is not currently registered to it
+            OperatorSet memory operatorSet = OperatorSet(params.avs, params.operatorSetIds[i]);
+            require(_operatorSets[operatorSet.avs].contains(operatorSet.id), InvalidOperatorSet());
+            require(!_isRegistered(operator, operatorSet), AlreadyMemberOfSet());
+
+            // Add operator to operator set
+            registeredSets[operator].add(operatorSet.key());
+            _operatorSetMembers[operatorSet.key()].add(operator);
+            emit OperatorAddedToOperatorSet(operator, operatorSet);
+
+            // Mark the operator registered
+            registrationStatus[operator][operatorSet.key()].registered = true;
+        }
+
+        // Call the AVS to complete registration. If the AVS reverts, registration will fail.
+        getAVSRegistrar(params.avs).registerOperator(operator, params.operatorSetIds, params.data);
+    }
+
+    /// @inheritdoc IAllocationManager
+    function deregisterFromOperatorSets(
+        DeregisterParams calldata params
+    ) external onlyWhenNotPaused(PAUSED_OPERATOR_SET_REGISTRATION_AND_DEREGISTRATION) {
+        // Check that the caller is either authorized on behalf of the operator or AVS
+        require(_checkCanCall(params.operator) || _checkCanCall(params.avs), InvalidCaller());
+
+        for (uint256 i = 0; i < params.operatorSetIds.length; i++) {
+            // Check the operator set exists and the operator is registered to it
+            OperatorSet memory operatorSet = OperatorSet(params.avs, params.operatorSetIds[i]);
+            require(_operatorSets[params.avs].contains(operatorSet.id), InvalidOperatorSet());
+            require(registrationStatus[params.operator][operatorSet.key()].registered, NotMemberOfSet());
+
+            // Remove operator from operator set
+            registeredSets[params.operator].remove(operatorSet.key());
+            _operatorSetMembers[operatorSet.key()].remove(params.operator);
+            emit OperatorRemovedFromOperatorSet(params.operator, operatorSet);
+
+            // Mark operator deregistered until the DEALLOCATION_DELAY passes
+            // forgefmt: disable-next-item
+            registrationStatus[params.operator][operatorSet.key()] = RegistrationStatus({
+                registered: false,
+                registeredUntil: uint32(block.number) + DEALLOCATION_DELAY
```

Comment:
i prefer to have this be called slashableUntil and modify other naming accordingly. registration should be synchronous.

keeping this called registered feels wrong when we modify all of the state and view function results at the point of time of this call...